python/lsst/daf/persistence/butler.py

```diff
@@ -136,44 +149,106 @@ def __init__(self, root, mapper=None, **mapperArgs):
+        if root is not None:
+            outputs = Repository.cfg(mode='rw',
+                                     storageCfg=PosixStorage.cfg(root=root), 
+                                     mapper=mapper, 
+                                     mapperArgs=mapperArgs)
 
-        :param cfg: a dict (or daf_persistence.Policy) of key-value pairs that describe the configuration.
-                    Best practice is to create this object by calling Butler.cfg()
-        :return: None
-        """
-        self._cfg = cfg
         self.datasetTypeAliasDict = {}
 
-        self.repository = Repository.makeFromCfg(self._cfg['repoCfg'])
-
         # Always use an empty Persistence policy until we can get rid of it
         persistencePolicy = pexPolicy.Policy()
         self.persistence = Persistence.getPersistence(persistencePolicy)
         self.log = pexLog.Log(pexLog.Log.getDefaultLog(), "daf.persistence.butler")
-
+        
+        if inputs is None:
+            inputs = []
+        elif not hasattr(inputs, '__iter__'):
+            inputs = [inputs]
+        
+        if outputs is None:
+            outputs = []
+        elif not hasattr(outputs, '__iter__'):
+            outputs = [outputs]
+        
+        self.inputs = []
+        self.outputs = [] 
+
+        for cfg in outputs:
+            if cfg['mode'] == 'r':
+                raise RuntimeError('Output repositories must not be read only.')
+            # All inputs become parents of each output, so add them to each output cfg:
+            cfg['parentCfgs'].extend(inputs)
+            # Outputs must not be read-only:
+            if cfg['mode'] == 'r':
+                raise RuntimeError('Output repositoires must not be read only.')
+            repoData = RepoData(cfg)
+            # readable outputs are also used as inputs:
+            self.outputs.append(repoData)
+            if cfg['mode'] == 'rw':
+                self._addInputs(repoData.cfg)
+
+        for cfg in inputs:
+            self._addInputs(cfg)
+
+        for repoData in self.outputs + self.inputs:
+            repoData.repo = Repository.makeFromCfg(repoData.cfg)
+
+    def _addInputs(self, cfg, tags=[]):
+        """Add a RepoData for cfg if it is not already represented in self.inputs."""
+        # Inputs must not be write-only:
+        if cfg['mode'] == 'w':
+            raise RuntimeError('Input repositoires must not be write only.')
+        repoData = None
+        tags = set(tags)
+        tags.update(set(cfg['tags']))
+
+        # If there is already an output RepoData instance for a given cfg, don't create a new one. Add it to
+        # the inputs list if it's not there, and update its tags.
+        for d in self.outputs:
+            if d.cfg == cfg:
+                self.inputs.append(d)
+                repoData = d
+
+        # If there is already an input RepoData instance for a given cfg, don't create a new one, just update 
+        # its tags.
+        if not repoData:
```

Comment:
I don't mean is not None though: e.g. "set() is None" is False, but "not set()" is True